### ISSUE TITLE
NTI misses "Property never defined" warning

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#992]
Using OTI (old type inference) I get this proper warning:

```
src/lab/model/CollisionAdvance.js:935: ERROR - Property impulse never defined on myphysicslab.lab.model.Collision
      return Math.max(r, c.impulse);
                           ^
```

on this code

``` javascript
/**
* @return {number}
* @private
*/
CollisionAdvance.prototype.maxImpulse = function() {
  return goog.array.reduce(this.collisions_,
    function(r, /** !myphysicslab.lab.model.Collision*/c, index, array) {
      return Math.max(r, c.impulse);
    }, 0);
};
```

And indeed the `Collision` interface does not define `impulse`.  Here is some of `Collision` interface:

``` javascript
@interface
*/
myphysicslab.lab.model.Collision = function() {};

/**
* @alias myphysicslab.lab.model.Collision
* @ignore
*/
var Collision = myphysicslab.lab.model.Collision;

/** 
@method
@return {number} the estimated time when this collision has `getDistance() == 0`,
or {NaN} if unknown
*/
Collision.prototype.getEstimatedTime;

/** 
@method
@return {number} the distance to the point where collision should be handled
*/
Collision.prototype.getDistance;
// code omitted...
```

But with NTI, there error is not detected.  I'm only getting warnings about "global this" for that file.

**Comment by dimvar** [comment]
That's pretty weird. Can you try to find a standalone testcase? It's hard for me to reproduce using the incomplete code snippets.

Also, try the following. Add this line as the first line in the function that uses c.impulse, so that we can see what the compiler thinks about the type of c.

    /** @type {null} */ var x = c;

You'll get a warning; copy the text here.

**Comment by myphysicslab** [comment]
Here is the error from adding that line:
```
src/lab/model/CollisionAdvance.js:523: WARNING - The right side in the assignment is not a
subtype of the left side.
left side  : null
right side : myphysicslab.lab.model.Collision

      /** @type {null} */ var x = c;
                                  ^
```
Here is the code with that line
```javascript
/**
* @return {number}
* @private
*/
CollisionAdvance.prototype.maxImpulse = function() {
  return goog.array.reduce(this.collisions_,
    function(r, /** !myphysicslab.lab.model.Collision*/c, index, array) {
      /** @type {null} */ var x = c;
      return Math.max(r, c.impulse);
    }, 0);
};
```
I'll try to find a small test case.

**Comment by myphysicslab** [comment]
I've made a test case including compile scripts for OTI and NTI.  A small zip archive is at
http://www.myphysicslab.com/rmnp/test20150609.zip
I show the code below as well.

file Collision.js
```javascript
goog.provide('myphysicslab.lab.Collision');

/** 
@interface
*/
myphysicslab.lab.Collision = function() {};

/**
@method
@return {number} the distance to the point where collision should be handled
*/
myphysicslab.lab.Collision.prototype.getDistance;
```

file CollisionAdvance.js
```javascript
goog.provide('myphysicslab.lab.CollisionAdvance');

goog.require('goog.array');
goog.require('goog.asserts');
goog.require('myphysicslab.lab.Collision');

goog.scope(function() {

var Collision = myphysicslab.lab.Collision;

/**
* @constructor
* @struct
*/
myphysicslab.lab.CollisionAdvance = function() {
  /**
  * @type {!Array.<!myphysicslab.lab.Collision>}
  */
  this.collisions_ = new Array();
};

var CollisionAdvance = myphysicslab.lab.CollisionAdvance;

/**
* @return {number}
*/
CollisionAdvance.prototype.maxImpulse = function() {
  return goog.array.reduce(this.collisions_,
    function(r, c, index, array) {
      return Math.max(r, c.foobar);
    }, 0);
};

CollisionAdvance.prototype.toString = function() {
  return "CollisionAdvance{"
    +"max impulse "+this.maxImpulse()
    +"}";
};

}); // goog.scope
```

OTI (old type inference) does give an error:
```
./lab/CollisionAdvance.js:30: ERROR - Property foobar never defined on myphysicslab.lab.Collision
      return Math.max(r, c.foobar);
                           ^
```

**Comment by dimvar** [comment]
Thanks for the example. So, in this case, does NTI find the warning if you annotate the type of c in the argument function? If yes, then this is a duplicate of #994.

**Comment by myphysicslab** [comment]
No warning from NTI with annotated type:
```javascript
/**
* @return {number}
*/
CollisionAdvance.prototype.maxImpulse = function() {
  return goog.array.reduce(this.collisions_,
    function(r, /** !myphysicslab.lab.Collision*/c, index, array) {
      return Math.max(r, c.foobar);
    }, 0);
};
```
I only get warnings on closure-library and externs, no warnings on my test code.

With OTI, get the same warning as before.

**Comment by myphysicslab** [comment]
I just retested and with NTI now getting the proper error with the annotated type:

```javascript
/**
* @return {number}
*/
CollisionAdvance.prototype.maxImpulse = function() {
  return goog.array.reduce(this.collisions_,
    function(r, /** !myphysicslab.lab.Collision*/c, index, array) {
      return Math.max(r, c.foobar);
    }, 0);
};
```
results in this warning:

```
./lab/CollisionAdvance.js:30: WARNING - Property foobar never defined on c of type 
myphysicslab.lab.Collision
      return Math.max(r, c.foobar);
                         ^
```

Without the annotated type still not getting that warning.

Compiler Version: v20151015-81-g69e1f5b

**Comment by myphysicslab** [comment]
This seems to be working without the annotated type now.  As far as I can tell this is all fixed.  My code for this is a little different now but I think this is similar:

```javascript
/** Returns the maximum impulse applied to any of the collisions.
* @param {!Array<!myphysicslab.lab.model.Collision>} collisions
* @return {number} maximum impulse applied
* @private
*/
CollisionAdvance.prototype.maxImpulse = function(collisions) {
  return goog.array.reduce(collisions, function(max, c, index, array) {
    var t = c.foo;
    var impulse = c.getImpulse();
    return isFinite(impulse) ? Math.max(max, impulse) : max;
  }, 0);
};
```

And I get the correct error with NTI:

```
src/lab/model/CollisionAdvance.js:859: WARNING - Property foo never defined on c of
type myphysicslab.lab.model.Collision
    var t = c.foo;
            ^^^^^
```
